Feature/mmjson parser

N283T opened the pull request:
## 📋 PR Checklist

- [ ] This PR is tagged as a [draft](https://github.blog/news-insights/product-news/introducing-draft-pull-requests/) if it is still under development and not ready for review. 
    > This avoids auto-triggering the slower tests in the CI and needlessly wasting resources.

- [x] I have ensured that all my commits follow [angular commit message conventions](https://www.conventionalcommits.org/en/v1.0.0-beta.4/).
    > Format: `<type>[optional scope]: <subject>`  
    > Example: `fix(af3): add missing crop transform to the af3 pipeline`
    >
    > This affects semantic versioning as follows:
    > - `fix`: patch version increment (0.0.1 → 0.0.2)
    > - `feat`: minor version increment (0.0.1 → 0.1.0) 
    > - `BREAKING CHANGE`: major version increment (0.0.1 → 1.0.0)
    > - All other types do not affect versioning
    >
    > The format ensures readable changelogs through auto-generation from commit messages.

- [x] I have run `make format` on the codebase before submitting the PR (this autoformats the code and lints it).

- [x] I have named the PR in angular PR message format as well (c.f. above), with a sensible tag line that summarizes all the changes in the PR. 
    > This is useful as the name of the PR is the default name of the commit that will be used if you merge with a squash & merge.
    > Format: `<type>[optional scope]: <subject>`  
    > Example: `fix(af3): add missing crop transform to the af3 pipeline`

---

## ℹ️ PR Description

### What changes were made and why?

This PR introduces support for parsing macromolecular JSON (mmJSON) files (see [documentation](https://pdbj.org/help/mmjson)), a format that is becoming increasingly popular for web-based structure visualization and transmission. 

Key changes include:
1.  **File Type Inference**: 
    - Updated `atomworks.io.utils.io_utils.infer_pdb_file_type` to correctly identify files with `.mmjson` or `.json` extensions as `mmjson` type.
    - Updated `CIF_LIKE_EXTENSIONS` to include `.mmjson` and `.mmjson.gz`.

2.  **Verification**:
    - Created a standalone verification script `mmjson_verification/verify_mmjson.py` to ensure correctness.
    - Added example data (`2hhb.json.gz` and `2hhb.cif.gz`) in `mmjson_verification/` to compare mmJSON parsing results against the standard CIF parser.
    - Verification script confirms that atom counts, coordinates, elements, residue names, and chain IDs match between the two formats.

---

Note: I am still new to contributing to this project, so any feedback or suggestions for improvement would be greatly appreciated. Thank you!

N283T committed:
feat: Add example CIF/JSON data, update dependencies, and modify I/O parsing utilities.

N283T committed:
feat: Add mmjson file type support, update file type inference, and introduce a parsing verification script with example data.

nscorley commented on mmjson_verification/2hhb.cif.gz:
Could we please move these files to `tests/data/io`? You'll see other similar test files there

nscorley commented on mmjson_verification/verify_mmjson.py:
In this project, we've been trying to use `pytest` - could we please convert this into a `pytest`-style test example, maybe in `tests/io/tools`?

nscorley commented on mmjson_verification/verify_mmjson.py:
```diff
@@ -0,0 +1,101 @@
+    from atomworks.io.parser import parse
+    from atomworks.io.utils.io_utils import infer_pdb_file_type
+
+PROJECT_ROOT = Path(__file__).parent.parent
```
If we correctly set up via `pytest`, the paths should be configured correctly already so we will be able to remove this

nscorley commented on mmjson_verification/verify_mmjson.py:
```diff
@@ -0,0 +1,101 @@
+    print("\n1. Testing File Type Inference...")
+    inferred_type = infer_pdb_file_type(json_path)
+    if inferred_type == "mmjson":
+        print(f"✅ Correctly inferred 'mmjson' from {json_path.name}")
```
When we convert to `pytest`, best-practice would be using `assert` rather than print statements

nscorley commented on mmjson_verification/verify_mmjson.py:
```diff
@@ -0,0 +1,101 @@
+        print("❌ Residue names mismatch.")
+        
+    # Compare Chain IDs
+    if np.array_equal(atoms_json.chain_id, atoms_cif.chain_id):
```
We have a utility for this - `assert_same_atom_array` - that should save quite a bit of code! We could just load into an `atom_array` with both methods and then use `assert_same_atom_array` which would be much more concise

nscorley commented on src/atomworks/io/utils/io_utils.py:
```diff
@@ -337,7 +351,16 @@ def infer_pdb_file_type(path_or_buffer: os.PathLike | io.StringIO | io.BytesIO)
         second_line = path_or_buffer.readline().strip()
         path_or_buffer.seek(0)
-        return "cif" if second_line.startswith("#") else "pdb"
+        path_or_buffer.seek(0)
```
The repeated seeks here are a little difficult to read -- maybe we just grab the first and second chars once then pattern match? perhaps something along the lines of:
```
def _infer_filetype(path_or_buffer: io.BytesIO | io.StringIO) -> str:
    """Infer file type from buffer contents."""
    if isinstance(path_or_buffer, io.BytesIO):
        return "bcif"
    
    # StringIO - peek at contents to determine format
    path_or_buffer.seek(0)
    first_char = path_or_buffer.read(1)
    path_or_buffer.readline()  # finish first line
    second_line = path_or_buffer.readline()
    path_or_buffer.seek(0)
    
    if first_char == "{":
        return "mmjson"
    if second_line.startswith("#"):
        return "cif"
    return "pdb"
    ```

nscorley commented on src/atomworks/io/utils/io_utils.py:
```diff
@@ -353,21 +376,40 @@ def infer_pdb_file_type(path_or_buffer: os.PathLike | io.StringIO | io.BytesIO)
+def _read_mmjson(file_obj: io.StringIO | io.BytesIO | io.TextIOWrapper) -> pdbx.CIFFile:
+    """Read an mmjson file into a CIFFile object."""
+    data = json.load(file_obj)
+    cif_file = pdbx.CIFFile()
```
nice job!

nscorley reviewed (changes requested):
Thank you for this thoughtful PR this is a great addition! A few comments on tests/style; it would also be worth running `make format` on the repository to clean up the formatting so the linter tests pass

N283T committed:
feat: add automated mmJSON parsing test and refactor buffer file type inference in `io_utils.py`.

N283T commented:
Thanks for the thoughtful review! I've addressed your feedback with the following changes:

- **Test Data**: Moved 2hhb.cif.gz and 2hhb.json.gz  to `tests/data/io` to align with the project structure.
- **Tests**: Converted the verification script into a proper pytest-style test in `tests/io/tools/test_mmjson.py`. I also switched to using `assert_same_atom_array` for more concise and robust assertions.
- **Refactoring**: Refactored `infer_pdb_file_type` in `io_utils.py` to `_infer_file_type_from_buffer` to improve readability and avoid repeated seeks.
- **Formatting**: Ran `make format` to ensure the code complies with the project's style guidelines.

Let me know if any additional changes are needed. Happy to adjust!

nscorley commented on tests/io/tools/test_mmjson.py:
```diff
@@ -0,0 +1,33 @@
+
+# Tests should look for data relative to their location or use a fixture
+# Based on project structure, tests/data is likely where we moved the files
+TEST_DATA_DIR = Path(__file__).parent.parent.parent / "data" / "io"
```
We have a fixture for this that we can import from `conftest` (as it looks like Claude suggested 😄, `TEST_DATA_DIR` )

nscorley commented on tests/io/tools/test_mmjson.py:
```diff
@@ -0,0 +1,33 @@
+
+    # 4. Compare Results
+    # This utility checks atom count, coordinates, and other annotations
+    assert_same_atom_array(atoms_json, atoms_cif)
```
Great!

nscorley reviewed (changes requested):
Looks good! Two minor changes:
1) Use the existing fixture for `TEST_DATA_DIR`
2) I think we still need to commit the test files - they might be being ignored so you need to `git add -f`

N283T committed:
test: add 2hhb mmJSON and CIF test data and update test to use shared data directory fixture

N283T commented:
Thanks for the review!

I've addressed the feedback:
- Updated `tests/io/tools/test_mmjson.py` to use the `TEST_DATA_DIR` fixture from `tests.conftest`.
- Force-added the test data files (`tests/data/io/2hhb.cif.gz` and `tests/data/io/2hhb.json.gz`) to git so they are no longer ignored.

(And yes — this time the AI assistant was Gemini Pro 😄)

nscorley commented:
Thanks! Looks great. I'm still seeing a linting error -- you can address with `make format` in the directory root; after that it'll be good to go!

N283T committed:
style: apply ruff formatting and lint fixes

N283T commented:
Ran `make format` from the repository root to resolve the remaining linting issue. Thanks!

N283T committed:
style: apply formatting fixes
